include/engine/plugins/plugin_base.hpp

```diff
@@ -371,6 +371,23 @@ class BasePlugin
         }
         return phantom_node_pairs;
     }
+
+    std::string MissingPhantomErrorMessage(const std::vector<PhantomNodePair> &phantom_nodes,
+                                           const std::vector<util::Coordinate> &coordinates) const
+    {
+        BOOST_ASSERT(phantom_nodes.size() < coordinates.size());
+        size_t missing_index = phantom_nodes.size();
+        for (size_t i : util::irange<size_t>(0, phantom_nodes.size()))
```

Comment:
Suggestion: I think this could be done with `std::mismatch`, which I think helps clarify intent a bit.

```cpp
auto mismatch = std::mismatch(phantom_nodes.begin(), phantom_nodes.end(), 
                              coordinates.begin(), coordinates.end(),
                              [](const auto &phantom_node, const auto &coordinate) {
                                return phantom_node.first.input_location == coordinate;
                              });
std::size_t missing_index = std::distance(phantom_nodes.begin(), mismatch.first);
```

Reply:
Nice. Yes, a lot clearer.